First iteration of TraversabilityGrid, TraversabilityCell and TraversabilityClass.

JanWehrmann opened the pull request:
Contains basic data structures and a serialization test.

saarnold commented on CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@
 cmake_minimum_required(VERSION 2.6)
 set(ROCK_USE_CXX11 TRUE)
 find_package(Rock)
+set(ROCK_TEST_ENABLED OFF)
```
This shouldn't be part of the PR.

saarnold commented on src/grid/TraversabilityClass.cpp:
```diff
@@ -0,0 +1,58 @@
+
+bool TraversabilityClass::isSet() const
+{
+  return drivability < 0;
```
The indentation seems to differ here. Please use everywhere (in this library) 4 spaces.

saarnold commented on src/grid/TraversabilityClass.hpp:
```diff
@@ -0,0 +1,91 @@
+	double getDrivability() const;
+	
+    private:
+	double drivability;
```
A float should have enough precision for the drivability.

saarnold commented on src/grid/TraversabilityGrid.hpp:
```diff
@@ -0,0 +1,116 @@
+      const TraversabilityClass& getTraversability(size_t x, size_t y) const;
+      
+      // Note that probability has to be in range [0,1].
+      void setProbability(double probability, size_t x, size_t y);
```
Since the probability is encoded in a uint8 a float would be enough here as well

saarnold commented on src/grid/TraversabilityGrid.hpp:
```diff
@@ -0,0 +1,116 @@
+            CLASS_OBSTACLE = 1,
+            CUSTOM_CLASSES = 2
+      };
+      CLASSES classes;
```
What is the reason for this member?

JanWehrmann replied:
It made serialization of itself possible, which is indeed unnecessary.

saarnold commented on src/grid/TraversabilityGrid.hpp:
```diff
@@ -0,0 +1,116 @@
+  class TraversabilityGrid : public GridMap<TraversabilityCell> 
+  {    
+  public:
+      enum CLASSES {
```
To align with the rest of the naming this should be CamelCase as well. Therefore just name it Classes.

saarnold commented on src/grid/TraversabilityGrid.hpp:
```diff
@@ -0,0 +1,116 @@
+      TraversabilityGrid();     
+      
+      // Use this if you want a map holding individual data.
+      TraversabilityGrid(const Vector2ui &num_cells, const Vector2d &resolution, const TraversabilityCell default_value);
```
default_value was supposed to be a reference right?

JanWehrmann replied:
Yes!

saarnold commented on test/serialization/test_serialization_Grid.cpp:
```diff
@@ -231,6 +232,57 @@ BOOST_AUTO_TEST_CASE(test_mlgrid_serialization)
+    {
+	for(unsigned j = 0; j < size.y(); j++)
+	{
+	    traversabilityGrid_out.setTraversabilityAndProbability(i % 4, j, i, j);
```
For the probability only values in [0,1] are valid right?

saarnold commented on test/serialization/test_serialization_Grid.cpp:
```diff
@@ -231,6 +232,57 @@ BOOST_AUTO_TEST_CASE(test_mlgrid_serialization)
+    
+    std::vector<TraversabilityClass> classes_out = traversabilityGrid_out.getTraversabilityClasses();
+    std::vector<TraversabilityClass> classes_in = traversabilityGrid_in.getTraversabilityClasses();
+    
```
Here you should also check if the girds and classes have the same size.

Commit:
First iteration of TraversabilityGrid, TraversabilityCell
and TraversabilityClass. Contains basic data structures
and a serialization test.

saarnold commented:
Ok looks good! Thanks!
